### PR TITLE
unl0kr: 2.0.0 -> 3.2.0

### PR DESCRIPTION
**Pull request by uninsane**
## Description of changes

- changelogs: <https://gitlab.com/postmarketOS/buffybox/-/releases>
  - new themes
  - UI fixes for the shutdown button
  - update vendored lvgl to 2023-06-03
- unl0kr has been moved to the buffybox umbrella project, [as announced in the old repo](https://gitlab.com/cherrypicker/unl0kr/-/commit/323629cfec302559b9c080a5a2c2eb0565fe7534)

## Things done

- Built on platform(s)
  - [x] x86_64-linux  (and `pkgCross.aarch64-multiplatform`, `pkgsCross.armv7l-hf-multiplatform`)
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by JohnRTitor** [comment]
Result of `nixpkgs-review pr 319126` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>unl0kr</li>
  </ul>
</details>

**Comment by JohnRTitor** [comment]
Ran the test (`nix-build -A nixosTests.systemd-initrd-luks-unl0kr`) locally and it hangs here.

```log

machine # [    1.924147] usbhid: USB HID core driver
machine # [    1.940566] input: QEMU QEMU USB Tablet as /devices/pci0000:00/0000:00:01.2/usb1/1-1/1-1:1.0/0003:0627:0001.0001/input/input2
machine # [    1.942567] hid-generic 0003:0627:0001.0001: input,hidraw0: USB HID v0.01 Mouse [QEMU QEMU USB Tablet] on usb-0000:00:01.2-1/input0
machine # [    2.379492] unl0kr-ask-password-start[116]: libinput error: /nix/store/279k8iw3mb4hdsqxsw22xdym4ayh03ss-libinput-1.25.0/share/libinput: failed to find data files
machine # [    2.380733] unl0kr-ask-password-start[116]: libinput error: Failed to load the device quirks from /nix/store/279k8iw3mb4hdsqxsw22xdym4ayh03ss-libinput-1.25.0/share/libinput and /etc/libinput/local-overrides.quirks. This will negatively affect device behavior. See https://wayland.freedesktop.org/libinput/doc/1.25.0/device-quirks.html for details.

```

**Review by tomfitzhenry — CHANGES_REQUESTED** [review]
Replying to @uninsane from https://github.com/NixOS/nixpkgs/pull/276539#issuecomment-2161531679 :

> see https://github.com/NixOS/nixpkgs/pull/319126 which is working for me on actual hardware (x86_64 laptop, aarch64 pinephone, armv7 chromebook). i tried to run the nix tests with nix-build -A nixosTests.systemd-initrd-luks-unl0kr, but nix says i need a builder with the kvm feature for that. pretty low on my own priorities to figure that one out and show that it works in a virtual environment when i already have it working on the device in my hand...

Glad to hear this is working on real hardware, and sorry I've not been proactive in doing this update.

One of the ways NixOS maintenance scales to its high packages-per-maintainer ratio is via its automated integration tests. It would be a step backwards in that dimension if we were to merge a PR that broke such a test. At the same time, I understand that having stale packages is itself not ideal.

I realise this puts us in a tricky spot. I'm no longer a user of unl0kr, but I see I'm the only maintainer. That's not good either! :)

Is there anyone willing to be a maintainer for unl0kr (package, module, and test)?

It'd be really great to us work with upstream to figure out why this test is failing. https://github.com/NixOS/nixpkgs/pull/276539#issuecomment-2051224051 was my past attempt, which isolated it to the 2.0.2 -> 2.0.3 upgrade. Maybe some git bisecting could find the commit that causes the test to break?

**Comment by hustlerone** [comment]
I am willing to maintain the unl0kr package since at least 2 of my devices use it.

**Comment by JohnRTitor** [comment]
Could you investigate the reason why the test is failing? It's a sensitive package after all, so we are hesitant on merging this.

**Comment by hustlerone** [comment]
I'm merging 2.0.3 before 3.2.0

I'll look into it

**Comment by JohnRTitor** [comment]
2.0.3 also fails the test (#276539) https://github.com/NixOS/nixpkgs/pull/276539/checks?check_run_id=19932461278

Alternatively we could merge this since it's tested by OP?

Testing this is relatively easy.

In `flake.nix`:

```nix
inputs.unlocker-test-nixpkgs.url = "github:NixOS/nixpkgs/refs/pull/319126/merge";
```

In `configuration.nix`:

```nix
  nixpkgs.overlays = [
    (final: prev: {
      unl0kr = inputs.unlocker-test-nixpkgs.legacyPackages.${pkgs.system}.unl0kr;
    })
  ];
```

**Comment by hustlerone** [comment]
I'll test it on a real install right now

**Comment by hustlerone** [comment]
It does NOT work. It hangs right before the moment the UI should load.

**Review by JohnRTitor — CHANGES_REQUESTED** [review]
Ok. let's block this, until this is fixed.

**Comment by Electrostasy** [comment]
I use an [overlay for 3.2.0](https://github.com/Electrostasy/dots/blob/30f93f117cfaadd82a68ab0c5532031e4cfc21eb/overlays/unl0kr_3_update.nix) since 3.0.0 without issues on 1 device (a 2022 Asus ROG Flow Z13). The only differences (from what I can tell) are that I do not fetch submodules and I do not specify `strictDeps = false`. Maybe this is relevant?

**Comment by hustlerone** [comment]
> I use an [overlay for 3.2.0](https://github.com/Electrostasy/dots/blob/30f93f117cfaadd82a68ab0c5532031e4cfc21eb/overlays/unl0kr_3_update.nix) since 3.0.0 without issues on 1 device (a 2022 Asus ROG Flow Z13). The only differences (from what I can tell) are that I do not fetch submodules and I do not specify `strictDeps = false`. Maybe this is relevant?

Only this? unl0kr does not execute on my system

**Comment by hustlerone** [comment]
Maybe I'm doing it wrong, but I added this as a module in my flake:

```
                                ({inputs, pkgs, ...}: {
                                        nixpkgs.overlays = [
                                                (final: prev: {
                                                  # Contains various fixes since 2.0.0.

                                                  unl0kr = prev.unl0kr.overrideAttrs (finalAttrs: oldAttrs: {
                                                    version = "3.2.0";

                                                    src = oldAttrs.src.override {
                                                      owner = "postmarketOS";
                                                      repo = "buffybox";
                                                      rev = finalAttrs.version;
                                                      hash = "sha256-nZX7mSY9IBIhVNmOD6mXI1IF2TgyKLc00a8ADAvVLB0=";
                                                    };

                                                    sourceRoot = "${finalAttrs.src.name}/unl0kr";
                                                  });
                                                })
                                        ];
                                }) # Overlays
```

**Comment by hustlerone** [comment]
> Maybe I'm doing it wrong, but I added this as a module in my flake:
> 
> ```
>                                 ({inputs, pkgs, ...}: {
>                                         nixpkgs.overlays = [
>                                                 (final: prev: {
>                                                   # Contains various fixes since 2.0.0.
> 
>                                                   unl0kr = prev.unl0kr.overrideAttrs (finalAttrs: oldAttrs: {
>                                                     version = "3.2.0";
> 
>                                                     src = oldAttrs.src.override {
>                                                       owner = "postmarketOS";
>                                                       repo = "buffybox";
>                                                       rev = finalAttrs.version;
>                                                       hash = "sha256-nZX7mSY9IBIhVNmOD6mXI1IF2TgyKLc00a8ADAvVLB0=";
>                                                     };
> 
>                                                     sourceRoot = "${finalAttrs.src.name}/unl0kr";
>                                                   });
>                                                 })
>                                         ];
>                                 }) # Overlays
> ```

Suspicions confirmed, unl0kr will NOT work if amdgpu is loaded.

**Comment by hustlerone** [comment]
> 2.0.3 also fails the test (#276539) https://github.com/NixOS/nixpkgs/pull/276539/checks?check_run_id=19932461278
> 
> Alternatively we could merge this since it's tested by OP?
> 
> Testing this is relatively easy.
> 
> In `flake.nix`:
> 
> ```nix
> inputs.unlocker-test-nixpkgs.url = "github:NixOS/nixpkgs/refs/pull/319126/merge";
> ```
> 
> In `configuration.nix`:
> 
> ```nix
>   nixpkgs.overlays = [
>     (final: prev: {
>       unl0kr = inputs.unlocker-test-nixpkgs.legacyPackages.${pkgs.system}.unl0kr;
>     })
>   ];
> ```
> 
> Note that you must have flakes setup and pass `inputs` via `specialArgs` to your NixOS configuration.

I'll retest this with amdgpu disabled

**Comment by hustlerone** [comment]
I'm going to see if DRM may be necessary

**Comment by hustlerone** [comment]
Looks like amdgpu has a conflict with unl0kr regardless.

**Comment by uninsane** [comment]
> Looks like amdgpu has a conflict with unl0kr regardless.

to be clear, 2.0.0 works for you, but 3.2.0 fails regardless _of `backend=fbdev`/`backend=drm` in `/etc/unl0kr.conf`?

and does your device show any signs of life in the display before reaching unl0kr (i.e. initrd logging)?

**Comment by hustlerone** [comment]
> > Looks like amdgpu has a conflict with unl0kr regardless.
> 
> to be clear, 2.0.0 works for you, but 3.2.0 fails regardless _of `backend=fbdev`/`backend=drm` in `/etc/unl0kr.conf`_?
> 
> and does your device show any signs of life in the display before reaching unl0kr (i.e. initrd logging)?

My device displays fine, it just frezes/unl0kr doesn't work.

2.0.3+ seems to fail regardless of the backend, and seems to be caused by amdgpu in stage 1

**Comment by hustlerone** [comment]
I'll probably step up as the maintainer since I want to add the much needed configuration feature (for themes and hacks).

FYI, having the config there is not enough, especially since unl0kr can't read it at boot (seems like a huge oversight, but then again NixOS has ways to work around this, which I've used and tested. All that remains is to translate my imperative crud to something declarative and stateful)

**Comment by Electrostasy** [comment]
> Looks like amdgpu has a conflict with unl0kr regardless.

I will add that I had a similar issue before with nvidia/i915 and early DRM, unl0kr would not start, so maybe this is not just a conflict with unl0kr and amdgpu. I have only tested this on 1 device, and it has hybrid graphics, so I was not sure if it was an isolated case. Interesting!

> I'll probably step up as the maintainer since I want to add the much needed configuration feature (for themes and hacks) and because I actively use unl0kr on at least two of my devices.
> 
> FYI, having the config there is not enough, especially since unl0kr can't read it at boot (seems like a huge oversight, but then again NixOS has ways to work around this, which I've used and tested. All that remains is to translate my imperative crud to something declarative and stateful)

I have been successfully using [a small module](https://github.com/Electrostasy/dots/blob/8877ae63da7c9aab676eef8c7fcf8f993a6ee84b/modules/unl0kr-settings/default.nix) to add config file support (used [like this](https://github.com/Electrostasy/dots/blob/3d16dab0bc12701e5a6c2bd00d6c8a72ff610b35/hosts/mercury/default.nix#L73-L82)), but adding extra commandline flags required more work. Having support for it in the official module would be huge.

**Comment by uninsane** [comment]
i use `unl0kr` on my aarch64 Pinephone: hardware which requires me to revert <https://github.com/NixOS/nixpkgs/pull/298332> (i.e. build the kernel with `CONFIG_FB_SIMPLE=y`) to get _any_ output in the initrd. since you're getting display output, and it's only unl0kr that doesn't render, that's _probably_ not the source of the issue, but it could be worth looking in that direction if there's not anything more obvious.

**Comment by hustlerone** [comment]
I'm first going to only update the service then attempt a version bump so we at least have an update on unl0kr

**Comment by hustlerone** [comment]
> i use `unl0kr` on my aarch64 Pinephone: hardware which requires me to revert #298332 (i.e. build the kernel with `CONFIG_FB_SIMPLE=y`) to get _any_ output in the initrd. since you're getting display output, and it's only unl0kr that doesn't render, that's _probably_ not the source of the issue, but it could be worth looking in that direction if there's not anything more obvious.

Just had unl0kr v2.0.0 freeze on my system. Looks like we really can't predict what'll happen. Either way, I'll do a PR with the changes done to the unl0kr service by [Electrostasy](https://github.com/Electrostasy)

**Comment by hustlerone** [comment]
Be warned that if it gets merged you might encounter a merge conflict

**Comment by uninsane** [comment]
i've switched my systems from unl0kr to buffyboard (a framebuffer keyboard, packaged [here](https://git.uninsane.org/colin/nix-files/src/commit/1e045c2c17120517a4cd1aa66edd3580bff7471b/pkgs/by-name/buffybox/package.nix)). leaving this PR open seems to have caused confusion elsewhere, that i intend to do more work on it (i don't), so i'm closing it. anyone who wants to shepherd this: feel free to pull the branch and open a new PR.

**Pull request closed without merging**
